redash/query_runner/couchbase.py

```diff
@@ -0,0 +1,172 @@
+                            "type": TYPES_MAP.get(type(row[key][inner_key]), TYPE_STRING)
+                        })
+
+                    parsed_row[column_name] = row[key][inner_key]
+
+            else:
+                if _get_column_by_name(columns, key) is None:
+                    columns.append({
+                        "name": key,
+                        "friendly_name": key,
+                        "type": TYPES_MAP.get(type(row[key]), TYPE_STRING)
+                    })
+
+                parsed_row[key] = row[key]
+
+        rows.append(parsed_row)
+    return rows, columns
+
+
+class Couchbase(BaseQueryRunner):
+
+    noop_query = 'Select 1'
+
+    @classmethod
+    def configuration_schema(cls):
+        return {
+            'type': 'object',
+            'properties': {
+                'host': {
+                    'type': 'string',
+                },
+                'port': {
+                    'type': 'string',
+                    'title': 'Port (Defaults: 8095 - Analytics, 8093 - N1QL)',
+                    'default': '8095'
+                },
+                'user': {
+                    'type': 'string',
+                },
+                'password': {
+                    'type': 'string',
+                },
+            },
+            'required': ['host', 'user', 'password'],
+            "order": ['host', 'port', 'user', 'password'],
+            "secret": ["password"]
+        }
+
+    def __init__(self, configuration):
+        super(Couchbase, self).__init__(configuration)
+
+    @classmethod
+    def enabled(cls):
+        return True
+
+    @classmethod
+    def annotate_query(cls):
+        return False
+
+    def test_connection(self):
+        result = self.call_service(self.noop_query, '')
+
+    def get_buckets(self, query, nameParam):
+        defaultColumns = [
+            'meta().id'
+        ]
+        result = self.call_service(query, "").json()['results']
+        schema = {}
+        for row in result:
+            table_name = row.get(nameParam)
+            schema[table_name] = {'name': table_name, 'columns': defaultColumns}
+
+        return schema.values()
+
+    def get_schema(self, get_stats=False):
+
+        try:
+            # Try fetch from Analytics
+            return self.get_buckets(
+                "SELECT ds.GroupName as name FROM Metadata.`Dataset` ds where ds.DataverseName <> 'Metadata'", "name")
+        except Exception:
+            # Try fetch from N1QL
+            return self.get_buckets("select name from system:keyspaces", "name")
+
+    def call_service(self, query, user):
+        try:
+            user = self.configuration.get("user")
+            password = self.configuration.get("password")
+            host = self.configuration.get("host")
+            port = self.configuration.get('port', 8095)
+            params = {'statement': query}
+
+            url = "http://%s:%s/query/service" % (host, port)
```

Comment:
Does Couchbase support HTTPS? In such case, we should probably make the protocol (http) configurable.

Reply:
extracted as parameter